src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1336 @@
+            that its nodata value is the global NODATA.
+        precip_array (numpy.ndarray): 2D array of precipitation amounts 
+            in millimeters/year
+        precip_nodata (float): nodata value for the precipitation array
+        pixel_area (float): area of each pixel in m^2
+    
+    Returns:
+        2D numpy.ndarray of precipitation volumes in m^3/year
+    """
+    volume_array = numpy.full(ratio_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (ratio_array != NODATA) & 
+        (precip_array != precip_nodata))
+
+    # precipitation (mm/yr) * pixel area (m^2) * 
+    # 0.001 (m/mm) * ratio = volume (m^3/yr)
+    volume_array[valid_mask] = (
+        precip_array[valid_mask] *
+        ratio_array[valid_mask] *
+        pixel_area * 0.001)
+    return volume_array
+
+
+def avoided_pollutant_load_op(lulc_array, lulc_nodata, retention_volume_array, 
+        sorted_lucodes, emc_array):
+    """Calculate avoided pollutant loads from LULC codes retention volumes.
+    This is intented to be used with pygeoprocessing.raster_calculator.
+
+    Args:
+        lulc_array (numpy.ndarray): 2D array of LULC codes
+        lulc_nodata (int): nodata value for the LULC array
+        retention_volume_array (numpy.ndarray): 2D array of stormwater 
+            retention volumes, with the same shape as ``lulc_array``. It is
+            assumed that the retention volume nodata value is the global NODATA
+        sorted_lucodes (numpy.ndarray): 1D array of the LULC codes in order
+            from smallest to largest
+        emc_array (numpy.ndarray): 1D array of pollutant EMC values for each 
+            lucode. ``emc_array[i]`` is the EMC for the LULC class at 
+            ``sorted_lucodes[i]``.
+
+    Returns:
+        2D numpy.ndarray with the same shape as ``lulc_array``.
+        Each value is the avoided pollutant load on that pixel in kg/yr,
+        or NODATA if any of the inputs have nodata on that pixel.
+    """
+    load_array = numpy.full(lulc_array.shape, NODATA, dtype=float)
+    valid_mask = (
+        (lulc_array != lulc_nodata) &
+        (retention_volume_array != NODATA))
+
+    # bin each value in the LULC array such that 
+    # lulc_array[i,j] == sorted_lucodes[lulc_index[i,j]]. thus, 
+    # emc_array[lulc_index[i,j]] is the EMC for the lucode at lulc_array[i,j]
+    lulc_index = numpy.digitize(lulc_array, sorted_lucodes, right=True)
+    # EMC for pollutant (mg/L) * 1000 (L/m^3) * 0.000001 (kg/mg) * 
+    # retention (m^3/yr) = pollutant load (kg/yr)
+    load_array[valid_mask] = (emc_array[lulc_index][valid_mask] * 
+        0.001 * retention_volume_array[valid_mask])
+    return load_array
+
+
+def retention_value_op(retention_volume_array, replacement_cost):
+    """Multiply array of retention volumes by the retention replacement 
+    cost to get an array of retention values. This is meant to be used with
+    raster_calculator.
+
```

Comment:
Kind of concerned about the validity of this valuation method... in my test data, using the recommended value of $1.59/cubic meter, with 100mm annual precipitation (not very much), retention on a roughly 40x60km polygon was valued at over $375,000,000. That seems pretty excessive. 
Maybe the valuation option should come with a disclaimer that it only makes sense when you run multiple scenarios and take the difference?

Reply:
I'm confident the math is right. When I send the finished model to Perrine and Ben, I'll ask about that.

Reply:
@emlys did we ever hear back from Ben and Perrine about this?

Reply:
Looking back through the email thread, Ben gave this response to my original question:
> Those big numbers are a bit concerning, but may be related to it being kind of the opposite of how these numbers are usually presented (most reporting deals with the export portion, which is typically a lot smaller and much easier to measure directly.)

I'll bring this up again in the user's guide PR or google doc, however we do that, and try to include a disclaimer there.